nixos/top-level: fail more helpfully on forbidden dependencies

lheckemann opened the pull request:
## Description of changes

The forbiddenDependenciesRegex option is useful for detecting unwanted dependencies, but not particularly helpful at discovering why they exist. This PR improves the situation using `nix why-depends` (in an awfully hacky, but working, way).

cc @nikstur with whom I came up with this.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

lheckemann committed:
nixos/top-level: fail more helpfully on forbidden dependencies

We now use nix why-depends in a terribly hacky way to show the paths
from the non-overridden toplevel to each forbidden dependency.

Note that this does not take into account dependencies from the build
closure, even if these are brought in using
system.includeBuildDependencies. I'm not sure if that's desirable or
not.

roberth commented:
> a more sophisticated solution

What about [system.forbiddenDependenciesRegex](https://search.nixos.org/options?channel=unstable&show=system.forbiddenDependenciesRegex&from=0&size=50&sort=relevance&type=packages&query=system.forbidden)?

I suppose we could improve the naming. "Dependencies" tends to be quite ambiguous - dependencies of what? The derivation or its outputs?
Also forbidden -> disallowed I guess, for discoverability.

roberth reviewed (changes requested):
This should most definitely come with tests, consider that this is rather complicated code that's easy to get wrong because of its many requirements.

lheckemann commented:
Indeed, I missed that! I'll try improving forbiddenDependenciesRegex instead.

lheckemann commented:
I've updated this to add helpful information for getting rid of forbidden dependencies, and added disallowedRequisites to the description for searchability (I don't think it's such a good idea to rename the option, since it's a very different mechanism).

Is there actually a good way to test build failures?

roberth commented:
@ofborg test activation

> Is there actually a good way to test build failures?

Actually `nixos/modules/system/activation/test.nix` has an example of `testBuildFailure`, but that function just turns a failing [`builder` (Nix)](https://nixos.org/manual/nix/stable/language/derivations#attr-builder) into one that succeeds and writes the log to `$out`. It can't flip Nix's built-in checks that run after realisation.
So it's great for testing the `testers`, but not for this case. Perhaps you could add a VM test?

lheckemann commented:
>It can't flip Nix's built-in checks that run after realisation.

That's fine in this case, since I've discarded the disallowedRequisites-based approach. Thanks!

lheckemann committed:
nixos/activation/test: check that the path to forbidden dependencies is shown

lheckemann commented:
@roberth In addition to moving away from the previous approach, I've added to the activation tests to ensure that the path-following actually works. What do you think?

Ma27 reviewed (approved):
LGTM, I like that change. Error output seems good to me as well :+1:

roberth reviewed (changes requested):
In the first commit message, you describe the change as terrible hacky, but not why that is, or what could be done about it. Is it about the incomplete store? Could you elaborate on that in the commit message, and/or take steps to improve the situation?
